File: server/src/main/kotlin/eu/mlich/lucy/model/finance/FinanceTransactionCategory.kt
package eu.mlich.lucy.model.finance

import java.util.*
import javax.persistence.*

/**
 * Represent a category of transactions.
 *
 * @author Juraj Mlich <jurajmlich@gmail.com>
 */
@Entity
@Table(name = "finance_transaction_category")
data class FinanceTransactionCategory(
        @Id
        @GeneratedValue(strategy = GenerationType.IDENTITY)
        @Column(name = "id")
        var id: Int? = null,

        @Column(name = "name")
        var name: String,

        @Column(name = "color")
        var color: String? = null,

        /**
         * Used primarily for expenses?
         */
        @Column(name = "negative")
        var negative: Boolean,

        @Column(name = "disabled")
        var disabled: Boolean,

        @Column(name = "public_key")
        var publicKey: UUID = UUID.randomUUID()

) {
    override fun equals(other: Any?) = other is FinanceDeposit && other.publicKey === publicKey

    override fun hashCode(): Int = publicKey.hashCode()
}

File: android/lib/model/finance_transaction_category.dart
import 'package:android/model/id_aware.dart';

class FinanceTransactionCategory extends IdAware<String> {
  String id;
  String name;
  String color;
  bool negative;
  bool disabled;

  FinanceTransactionCategory(this.id);

  @override
  String toString() {
    return 'FinanceTransaction{id: $id, name: $name, color: $color, negative: '
        '$negative, disabled: $disabled}';
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is FinanceTransactionCategory &&
          runtimeType == other.runtimeType &&
          id == other.id;

  @override
  int get hashCode => id.hashCode;
}
